### ISSUE TITLE
performance questions

### ISSUE THREAD

**Issue by bf31415** [linkedin/goavro#69]
I'm using this goavro package with kafka and seeing less than hoped for performance.

In my setup, if I pull just the kafka messages and do not call the avro decode function, my process can consume 200k msgs/sec.  This is an average over multiple millions of messages.   each message is about 130 bytes.   my schema has 21 fields (a mix of fixed, long and string) and if I attempt to decode all 21 fields my processing msgs/sec falls to about 35k msgs/sec (again averaged over millions of msgs).   Ouch.  

if I decode the first five fields, the code can process about 115k msgs/sec.  

basically as I increase the number of the schema fields to be decoded, I'm seeing reduced performance.  Ok, i get this but i'm surprised by the performance hit.

is this expected?

i'm using the code in the decodeRecord example namely:

       encoded :=  // message from kafka
	bb := bytes.NewBuffer(encoded)
	decoded, err := codec.Decode(bb)

these measurements don't include any procesesing beyond the code.Decode(bb) line above.

Are there ways to improve the performance I'm seeing from the native avro decode mechanism? 

Thanks

**Comment by karrick** [comment]
@bf31415, could you upload a sanitized version of the schema you're using?

I'd like to run a few benchmarks locally.

**Comment by karrick** [comment]
@bf31415,

I'm still happy to run some benchmarks with a sanitized version of that schema. In the meanwhile, please take a look at the newer version of this library at https://github.com/karrick/goavro

On hefty payloads at LinkedIn, it has shown itself to be 3x encoding performance, and 4x decoding performance. YMMV.

Another thing to keep in mind is that both versions of the goavro library, this one and the one linked above, mutate no state after instantiation. You can have many running in parallel on the same box to decode more messages in parallel. Provided your particular application does not have an absolute strict ordering requirement for messages, running many of these decoders in worker pools would greatly speed up decoding. That's what some of our applications do.

I'm closing this issue as "Won't Fix" because performance was drastically improved in the replacement library.
